plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesFunctionProcessor.java

```diff
@@ -0,0 +1,209 @@
+    }
+
+    private TableFunctionProcessorState processCdfFile()
+    {
+        Page page = deltaLakePageSource.getNextPage();
+        if (page != null) {
+            int passThroughColumns = splitColumns.size();
+            Block[] resultBlock = new Block[passThroughColumns + NUMBER_OF_ADDITIONAL_COLUMNS_FOR_CDF_FILE];
+            for (int i = 0; i < splitColumns.size(); i++) {
+                resultBlock[i] = page.getBlock(i);
+            }
+            resultBlock[passThroughColumns] = RunLengthEncodedBlock.create(
+                    nativeValueToBlock(INTEGER, version), page.getPositionCount());
+            long commitTimestampMillis = Instant.ofEpochMilli(commitTimestamp).toEpochMilli();
+            resultBlock[passThroughColumns + 1] = RunLengthEncodedBlock.create(
+                    nativeValueToBlock(TIMESTAMP_TZ_MILLIS, packDateTimeWithZone(commitTimestampMillis, UTC_KEY)), page.getPositionCount());
+            return TableFunctionProcessorState.Processed.usedInputAndProduced(new Page(page.getPositionCount(), resultBlock));
+        }
+        return FINISHED;
+    }
+
+    private TableFunctionProcessorState processDataFile()
+    {
+        Page page = deltaLakePageSource.getNextPage();
+        if (page != null) {
+            int passThroughColumns = splitColumns.size();
+            Block[] resultBlock = new Block[passThroughColumns + NUMBER_OF_ADDITIONAL_COLUMNS_FOR_DATA_FILE];
+            for (int i = 0; i < splitColumns.size(); i++) {
+                resultBlock[i] = page.getBlock(i);
+            }
+            resultBlock[passThroughColumns] = RunLengthEncodedBlock.create(
+                    nativeValueToBlock(VARCHAR, utf8Slice("insert")), page.getPositionCount());
+            resultBlock[passThroughColumns + 1] = RunLengthEncodedBlock.create(
+                    nativeValueToBlock(INTEGER, version), page.getPositionCount());
+            long commitTimestampMillis = Instant.ofEpochMilli(commitTimestamp).toEpochMilli();
+            resultBlock[passThroughColumns + 2] = RunLengthEncodedBlock.create(
+                    nativeValueToBlock(TIMESTAMP_TZ_MILLIS, packDateTimeWithZone(commitTimestampMillis, UTC_KEY)), page.getPositionCount());
+            return TableFunctionProcessorState.Processed.usedInputAndProduced(new Page(page.getPositionCount(), resultBlock));
+        }
+        return FINISHED;
+    }
+
+    private DeltaLakePageSource createDeltaLakePageSource(ConnectorSession session, TableChangesSplit split)
+    {
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        TrinoInputFile inputFile = fileSystem.newInputFile(split.getPath());
+        Map<String, Optional<String>> partitionKeys = split.getPartitionKeys();
+        List<String> partitionValues = new ArrayList<>();
+        for (DeltaLakeColumnHandle column : splitColumns) {
+            Optional<String> value = partitionKeys.get(column.getName());
+            if (value != null) {
```

Comment:
Is the `if (value != null) {` necessary?

Reply:
It is, if there is null we don't add it, if the value of partition is null we add it

Reply:
we do same thing here https://github.com/trinodb/trino/blob/5150eb5045d1a9a601e7d485038f41006109f782/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSourceProvider.java#L136